crates/matrix-sdk-store-encryption/src/lib.rs

```diff
@@ -182,41 +264,66 @@ impl StoreCipher {
     /// # anyhow::Ok(()) };
     /// ```
     pub fn import(passphrase: &str, encrypted: &[u8]) -> Result<Self, Error> {
-        let encrypted: EncryptedStoreCipher = serde_json::from_slice(encrypted)?;
+        // Our old export format used serde_json for the serialization format, let's
+        // first try the new format, if that fails try the old one.
+        let encrypted: EncryptedStoreCipher =
+            if let Ok(deserialized) = rmp_serde::from_slice(encrypted) {
+                deserialized
+            } else {
+                serde_json::from_slice(encrypted)?
+            };
 
         let key = match encrypted.kdf_info {
             KdfInfo::Pbkdf2ToChaCha20Poly1305 { rounds, kdf_salt } => {
                 Self::expand_key(passphrase, &kdf_salt, rounds)
             }
-        };
-
-        let key = ChachaKey::from_slice(key.as_ref());
-
-        let mut decrypted = match encrypted.ciphertext_info {
-            CipherTextInfo::ChaCha20Poly1305 { nonce, ciphertext } => {
-                let cipher = XChaCha20Poly1305::new(key);
-                let nonce = XNonce::from_slice(&nonce);
-                cipher.decrypt(nonce, ciphertext.as_ref())?
+            KdfInfo::None => {
+                return Err(Error::KdfMismatch);
             }
         };
 
-        if decrypted.len() != 64 {
-            decrypted.zeroize();
+        let key = ChachaKey::from_slice(key.as_ref());
 
-            Err(Error::Length(64, decrypted.len()))
-        } else {
-            let mut encryption_key = Box::new([0u8; 32]);
-            let mut mac_key_seed = Box::new([0u8; 32]);
+        Self::import_helper(key, encrypted)
+    }
 
-            encryption_key.copy_from_slice(&decrypted[0..32]);
-            mac_key_seed.copy_from_slice(&decrypted[32..64]);
+    /// Restore a store cipher from an encrypted export.
```

Comment:
```suggestion
    /// Restore a store cipher from an export encrypted with a random key.
```

The `import` method above should probably receive the same treatment and say something like

```
    /// Restore a store cipher from an export encrypted with a passphrase.